### PR TITLE
Closes #6186: Add app version to Socorro crash report

### LINKED ISSUE #6186: Use app version for Socorro crash reports
Currently GeckoView version is used for Socorro crash reports.  The correct version number should be the app version.

### PR DESCRIPTION
**Pull request by rocketsroger**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 1]
```diff
@@ -55,11 +62,27 @@ class MozillaSocorroService(
     private val version: String = BuildConfig.MOZILLA_VERSION,
     private val buildId: String = BuildConfig.MOZ_APP_BUILDID,
     private val vendor: String = BuildConfig.MOZ_APP_VENDOR,
-    private val serverUrl: String = "https://crash-reports.mozilla.com/submit?id=$appId&version=$version&$buildId"
+    private var serverUrl: String? = null,
+    private var versionName: String = "0.0"
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Why can this get passed in if we override it in `init()`?

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 2]
```diff
@@ -55,11 +62,27 @@ class MozillaSocorroService(
...
+                .getPackageInfo(applicationContext.packageName, 0).versionName
+        } catch (e: PackageManager.NameNotFoundException) {
+            Logger.error("package name not found, failed to get application version")
+        } catch (e: IllegalStateException) {
+            Logger.error("failed to get application version")
+        }
+
+        if (serverUrl == null) {
+            serverUrl = "https://crash-reports.mozilla.com/submit?id=$appId&version=$versionName"
+        }
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: How will that work on crash-stats which seems to only show Gecko versions?

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 3]
```diff
@@ -123,7 +146,11 @@ class MozillaSocorroService(
         val gzipOs = GZIPOutputStream(os)
         sendPart(gzipOs, boundary, "ProductName", appName, nameSet)
         sendPart(gzipOs, boundary, "ProductID", appId, nameSet)
-        sendPart(gzipOs, boundary, "Version", version, nameSet)
+        sendPart(gzipOs, boundary, "Version", versionName, nameSet)
+        sendPart(gzipOs, boundary, "AcVersion", AcBuild.version, nameSet)
+        sendPart(gzipOs, boundary, "GleanVersion", AcBuild.gleanSdkVersion, nameSet)
+        sendPart(gzipOs, boundary, "AsVersion", AcBuild.applicationServicesVersion, nameSet)
+        sendPart(gzipOs, boundary, "GeckoViewVersion", version, nameSet)
```
Comment (pocmo) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Will those other versions get exposed in the crash-stats UI? :)

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 2 above)
Comment (willkg) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: The collector ignores querystring parameters in the crash report submission, but they sometimes help people to debug issues.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 3 above)
Comment (willkg) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Crash Stats shows fields that are explicitly marked public to everyone. Fields that aren't explicitly marked public only show up for people who have access to PII.

If you want additional fields to show up for everyone, be searchable or aggregatable, or something along those lines, please open up a bug against the Socorro product and we can work through the request.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 4]
```diff
@@ -123,7 +146,11 @@ class MozillaSocorroService(
         val gzipOs = GZIPOutputStream(os)
         sendPart(gzipOs, boundary, "ProductName", appName, nameSet)
         sendPart(gzipOs, boundary, "ProductID", appId, nameSet)
-        sendPart(gzipOs, boundary, "Version", version, nameSet)
+        sendPart(gzipOs, boundary, "Version", versionName, nameSet)
+        sendPart(gzipOs, boundary, "AcVersion", AcBuild.version, nameSet)
+        sendPart(gzipOs, boundary, "GleanVersion", AcBuild.gleanSdkVersion, nameSet)
+        sendPart(gzipOs, boundary, "AsVersion", AcBuild.applicationServicesVersion, nameSet)
```
Comment (willkg) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: I would spell this out and not abbreviate it.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt [hunk 5]
```diff
@@ -123,7 +146,11 @@ class MozillaSocorroService(
         val gzipOs = GZIPOutputStream(os)
         sendPart(gzipOs, boundary, "ProductName", appName, nameSet)
         sendPart(gzipOs, boundary, "ProductID", appId, nameSet)
-        sendPart(gzipOs, boundary, "Version", version, nameSet)
+        sendPart(gzipOs, boundary, "Version", versionName, nameSet)
+        sendPart(gzipOs, boundary, "AcVersion", AcBuild.version, nameSet)
```
Comment (willkg) [inline components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: I would spell this out and not abbreviate it.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 1 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: I gave it a default value for the case when the caller didn't pass in a value or we failed to get the `PackageInfo`.
This way we will at least have "0.0" as the value.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 5 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Understood.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 4 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Agreed.

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 1 above)
Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: After some thoughts, I've updated this to default as "N/A".  This makes more sense when we fails to (shouldn't happen) retrieve `PackageInfo`.

**Comment by rocketsroger** [comment]
Request for data collection review
---
1. What questions will you answer with this data?
            This data will give us the following data when a crash occurs: GeckoView version, Android components version, Glean version, Application Services version, application version and crash type.
 
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements?
            Yes, this allow us to know which specific build this crash came from so it makes it easier to reproduce the crash.
 
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
            No alternative is available to answer this question.
 
4. Can current instrumentation answer these questions?
            No, current instrumentation only tells us the GeckoView version.
 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki.

Measurement   Description | Data Collection   Category | Tracking Bug #
-- | -- | --
App version | category 1 - technical   data | https://github.com/mozilla-mobile/android-components/issues/6186
Android Component version | category 1 - technical   data | https://github.com/mozilla-mobile/android-components/issues/6186
Glean version | category 1 - technical   data | https://github.com/mozilla-mobile/android-components/issues/6186
Application Services version | category 1 - technical   data | https://github.com/mozilla-mobile/android-components/issues/6186
GeckoView version | category 1 - technical   data | https://github.com/mozilla-mobile/android-components/issues/6186
Crash type | category 1 - technical   data | https://github.com/mozilla-mobile/android-components/issues/6186

6. How long will this data be collected? Choose one of the following:
            I want to permanently monitor this data. (Roger Yang)
 
7. What populations will you measure?
            No filters - all locales, channels, etc.
 
8. If this data collection is default on, what is the opt-out mechanism for users?
            User will be prompted before collection.  User can opt-out using the prompt.
 
9. Please provide a general description of how you will analyze this data.
            With crash report, we will use the included version info to reproduce the issue.
 
10. Where do you intend to share the results of your analysis?
            Internally in the fenix/android-components teams.
 
11. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
            No third-party tool will use this data.

**Comment by rocketsroger** [comment]
@liuche Please review the request for data collection. Thanks!

File: components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt (hunk 2 above)
Comment (pocmo) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Nice. In this case maybe consider adding the AC version additionally (which is maybe more interesting than the app version if we need to debug something regarding crash reports), @rocketsroger?

Comment (rocketsroger) [reply components/lib/crash/src/main/java/mozilla/components/lib/crash/service/MozillaSocorroService.kt]: Agreed.  Updated `serverUrl`

**Comment by mmccorks** [comment]
For the data review: Are these new fields documented anywhere? (similar to the probe dictionary on desktop or the telemetry.md docs for focus and fenix?)

**Comment by rocketsroger** [comment]
> For the data review: Are these new fields documented anywhere? (similar to the probe dictionary on desktop or the telemetry.md docs for focus and fenix?)

My mistake.  I'll update the change.  Thanks,

**Comment by rocketsroger** [comment]
> For the data review: Are these new fields documented anywhere? (similar to the probe dictionary on desktop or the telemetry.md docs for focus and fenix?)

Updated.  Added what version information that will be collected in the README.md

**Comment by mmccorks** [comment]
**Data Review Form**
1.    Is there or will there be documentation that describes the schema for the ultimate data set in a public, complete, and accurate way? 
Yes, this is documented in the README and the Glean docs. 

2.    Is there a control mechanism that allows the user to turn the data collection on and off?
Yes, users can opt out of telemetry collection. 

3.    If the request is for permanent data collection, is there someone who will monitor the data over time?
Data will be monitored by @rocketsroger. 

4.    Using the category system of data types on the Mozilla wiki, what collection type of data do the requested measurements fall under?
Category 1, Technical Data.

5.    Is the data collection request for default-on or default-off?
Default on. 

6.    Does the instrumentation include the addition of any new identifiers?
No new identifiers.

7.    Is the data collection covered by the existing Firefox privacy notice? 
Yes.

8.    Does there need to be a check-in in the future to determine whether to renew the data? 
No check in needed. 

9.    Does the data collection use a third-party collection tool? 
No. 

----- 
data-review +

**Comment by rocketsroger** [comment]
> @rocketsroger should there also be an update to https://github.com/mozilla-mobile/android-components/blob/master/components/lib/crash/metrics.yaml ?

Unfortunately that is only for Glean.  I do agree that we should have a central location to track extra information that is being sent to the Socorro server.  I've created a issue to track this:
https://github.com/mozilla-mobile/android-components/issues/6223

Thanks,

**Comment by rocketsroger** [comment]
bors r=pocmo

**Comment by ebalazs-sv** [comment]
The issue https://github.com/mozilla-mobile/android-components/issues/6186 is verified as fixed, so we can also move this to Done. Thanks!

**Pull request merged**